developer/src/kmcmplib/tests/gtest-compiler-test.cpp

```diff
@@ -0,0 +1,428 @@
+// int GetCompileTargetsFromTargetsStore(const KMX_WCHAR* store)
+
+TEST_F(CompilerTest, IsValidKeyboardVersion_test) {
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u""));
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u" "));
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u"\t"));
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u" 1.1"));    
+    EXPECT_TRUE(IsValidKeyboardVersion((KMX_WCHAR *)u"1.1"));
+    EXPECT_TRUE(IsValidKeyboardVersion((KMX_WCHAR *)u"1.0"));
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u"1."));
+    EXPECT_TRUE(IsValidKeyboardVersion((KMX_WCHAR *)u"1.2.3"));
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u"a"));
+    EXPECT_FALSE(IsValidKeyboardVersion((KMX_WCHAR *)u"1.a"));
+};
+
+// KMX_DWORD kmcmp::AddCompilerVersionStore(PFILE_KEYBOARD fk)
+// KMX_DWORD CheckStatementOffsets(PFILE_KEYBOARD fk, PFILE_GROUP gp, PKMX_WCHAR context, PKMX_WCHAR output, PKMX_WCHAR key)
+// KMX_BOOL CheckContextStatementPositions(PKMX_WCHAR context)
+// KMX_DWORD CheckUseStatementsInOutput(PKMX_WCHAR output)
+// KMX_DWORD CheckVirtualKeysInOutput(PKMX_WCHAR output)
+// KMX_DWORD InjectContextToReadonlyOutput(PKMX_WCHAR pklOut)
+// KMX_DWORD CheckOutputIsReadonly(const PFILE_KEYBOARD fk, const PKMX_WCHAR output)
+// KMX_DWORD ProcessKeyLine(PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX_BOOL IsUnicode)
+// KMX_DWORD ProcessKeyLineImpl(PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX_BOOL IsUnicode, PKMX_WCHAR pklIn, PKMX_WCHAR pklKey, PKMX_WCHAR pklOut)
+// KMX_DWORD ExpandKp_ReplaceIndex(PFILE_KEYBOARD fk, PFILE_KEY k, KMX_DWORD keyIndex, int nAnyIndex)
+// KMX_DWORD ExpandKp(PFILE_KEYBOARD fk, PFILE_KEY kpp, KMX_DWORD storeIndex)
+// PKMX_WCHAR GetDelimitedString(PKMX_WCHAR *p, KMX_WCHAR const * Delimiters, KMX_WORD Flags)
+// LinePrefixType GetLinePrefixType(PKMX_WCHAR *p)
+// int LineTokenType(PKMX_WCHAR *str)
+// KMX_BOOL StrValidChrs(PKMX_WCHAR q, KMX_WCHAR const * chrs)
+// KMX_DWORD GetXString(PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX_WCHAR const * token,
+//  PKMX_WCHAR output, int max, int offset, PKMX_WCHAR *newp, int /*isVKey*/, int isUnicode
+// )
+
+TEST_F(CompilerTest, GetXStringImpl_test) {
+    KMX_WCHAR tstr[128];
+    FILE_KEYBOARD fk;
+    KMX_WCHAR str[LINESIZE];
+    KMX_WCHAR output[GLOBAL_BUFSIZE];
+    PKMX_WCHAR newp = NULL;
+
+    // CERR_BufferOverflow, max=0
+    EXPECT_EQ(CERR_BufferOverflow, GetXStringImpl(tstr, &fk, str, u"", output, 0, 0, &newp, FALSE));
+
+    // CERR_None, no token
+    str[0] = '\0';
+    EXPECT_EQ(CERR_None, GetXStringImpl(tstr, &fk, str, u"", output, 80, 0, &newp, FALSE));
+
+    // CERR_NoTokensFound, empty
+    u16cpy(str, u"");
+    EXPECT_EQ(CERR_NoTokensFound, GetXStringImpl(tstr, &fk, str, u"c", output, 80, 0, &newp, FALSE));
+
+    // CERR_NoTokensFound, whitespace
+    u16cpy(str, u" ");
+    EXPECT_EQ(CERR_NoTokensFound, GetXStringImpl(tstr, &fk, str, u"c", output, 80, 0, &newp, FALSE));
+}
+
+TEST_F(CompilerTest, GetXStringImpl_type0_test) {
```

Comment:
I wonder if we should name these according to the prefix char:

```suggestion
// tests strings starting with 'x' or 'd'
TEST_F(CompilerTest, GetXStringImpl_type_xd_test) {
```

Reply:
Good suggestion, especially as I want to elimiate the type literals from the source in due course. Done